src/com/google/javascript/jscomp/ProductionCoverageInstrumentationCallback.java

```diff
@@ -0,0 +1,188 @@
+   * This constant represents what the INSTRUMENT_CODE_FILE_NAME module will be renamed to by the
+   * compiler and this will be used to make the correct call to INSTRUMENT_CODE_FUNCTION_NAME.
+   */
+  private static final String MODULE_RENAMING = "module$exports$instrument$code";
+
+  /**
+   * INSTRUMENT_CODE_FILE_NAME will contain an instance of the instrumentCode class and that
+   * instance name is stored in this constant.
+   */
+  private static final String INSTRUMENT_CODE_INSTANCE = "instrumentCodeInstance";
+
+  private final AbstractCompiler compiler;
+  private final ParameterMapping parameterMapping;
+  boolean visitedInstrumentCodeFile = false;
+  private String functionName = "Anonymous";
+
+  public ProductionCoverageInstrumentationCallback(
+      AbstractCompiler compiler) {
+    this.compiler = compiler;
+    this.parameterMapping = new ParameterMapping();
+  }
+
+  @Override
+  public void visit(NodeTraversal traversal, Node node, Node parent) {
+    String fileName = traversal.getSourceName();
+    String sourceFileName = node.getSourceFileName();
+
+    // If origin of node is not from sourceFile, do not instrument. This typically occurs when
+    // polyfill code is injected into the sourceFile AST and this check avoids instrumenting it. We
+    // avoid instrumentation as this callback does not distinguish between sourceFile code and
+    // injected code and can result in an error.
+    if (!Objects.equals(fileName, sourceFileName)) {
+      return;
+    }
+
+    // If Source File INSTRUMENT_CODE_FILE_NAME has not yet been visited, do not instrument as
+    // the instrument function has not yet been defined and any call made to it will result in an
+    // error in the compiled JS code.
+    if (!visitedInstrumentCodeFile || sourceFileName.endsWith(INSTRUMENT_CODE_FILE_NAME)) {
+      if (sourceFileName.endsWith(INSTRUMENT_CODE_FILE_NAME)) {
+        visitedInstrumentCodeFile = true;
+      }
+      return;
+    }
+
+    if (node.isFunction()) {
+      functionName = NodeUtil.getBestLValueName(NodeUtil.getBestLValue(node));
+      instrumentCode(traversal, node.getLastChild(), functionName);
+    }
+  }
+
+  /**
+   * Iterate over all collected block nodes within a Script node and add a new child to the front of
+   * each block node which is the instrumentation Node
+   *
+   * @param traversal The node traversal context which maintains information such as fileName being
+   *                  traversed
+   * @param block     The block node to be instrumented instrumented
+   * @param fnName    The function name that encapsulates the current node block being instrumented
+   */
+  private void instrumentCode(NodeTraversal traversal, Node block, String fnName) {
+    block.addChildToFront(
+        newInstrumentationNode(traversal, block, fnName));
+    compiler.reportChangeToEnclosingScope(block);
+  }
+
+  /**
+   * Create a function call to the Instrument Code function with properly encoded parameters.
+   *
```

Comment:
Can you add example of how that function call looks like so that it's easier to understand structure of the node you are building.

Reply:
Added two examples, one with the constants and another with what the current constants are to. We can remove the example of what the current constants evaluate to when we make them dynamic.